wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -145,35 +158,69 @@ public void validate(Environment environment)
    * @param workspace the workspace to inject runtime into
    */
   public void injectRuntime(WorkspaceImpl workspace) throws ServerException {
-    RuntimeState runtimeState = runtimes.get(workspace.getId());
-    if (runtimeState != null) {
+    try (Unlocker ignored = lockService.writeLock(workspace.getId())) {
+      WorkspaceStatus workspaceStatus = statuses.get(workspace.getId());
+
+      if (workspaceStatus == null) {
+        workspace.setStatus(STOPPED);
+        return;
+      }
+
       try {
-        workspace.setRuntime(asRuntime(runtimeState));
-      } catch (InfrastructureException e) {
-        throw new ServerException(
-            "Error occurred while runtime state inspection. " + e.getMessage());
+        workspace.setRuntime(asRuntime(getRuntime(workspace.getId())));
```

Comment:
`asRuntime(getRuntime())` doesn't seem very clear to me. Maybe `asRuntime(getInternalRuntime())`?